### ISSUE TITLE
mypy fails with partial stubs and multiple files

### ISSUE THREAD

**Issue by sharmaeklavya2** [python/mypy#1550]
The module `six.moves.configparser` is not in typeshed, but some parts of `six.moves` (like `six.moves.range`) are in typeshed.

Consider these 2 files:
- file1.py:

``` py
import six.moves.configparser
config_file = six.moves.configparser.RawConfigParser()
```
- file2.py

``` py
from six.moves import range

for i in range(4):
    print(i)
```

These files pass mypy individually (i.e. `mypy --silent-import file1.py` and `mypy --silent-import file2.py` don't give any errors). But they fail when checked together:

```
$ mypy --silent-import file1.py file2.py
file1.py:2: error: "module" has no attribute "moves"
```

**Comment by gvanrossum** [comment]
Seems unrelated to `--silent-imports`. Without that I get the same error:
```
$ my file1.py file2.py 
file1.py:1: error: No library stub file for module 'six.moves.configparser'
file1.py:1: note: (Stub files are from https://github.com/python/typeshed)
file1.py:2: error: "module" has no attribute "moves"
```
This does look like something's wrong.

**Comment by ddfisher** [comment]
It seems likely that this is related to an import cycle in six.  Needs further investigation.

**Comment by gvanrossum** [comment]
Hm, `mypy -v file1.py file2.py` does not bear out that theory. It does not report any import cycles involving six. I have another theory, which is that there's something wrong with the way `from ... import ...` is processed. But I really don't want to spend more time on this right now.

**Comment by li-dan** [comment]
It's possible to work around this bug by changing the stubs for `six.moves`. `six.moves.configparser` is currently implemented with a line inside the stub for `six.moves`. On Python 2, from https://github.com/python/typeshed/blob/master/third_party/2/six/moves/__init__.pyi:

```python
import ConfigParser as configparser
```

The bug goes away if we instead create a stub file `third_party/2/six/moves/configparser.pyi` containing:

```python
from ConfigParser import *
```

I propose doing this for all moved modules in `six.moves`. (Note that this bug affects Python 2 and 3, so we'd apply the same workaround for both.)

One potential problem is whether it's okay to modify typeshed in order to work around a mypy bug.

**Comment by gvanrossum** [comment]
I think this is a good approach. Are you interested in submitting a PR?

**Comment by li-dan** [comment]
Sure, I'll submit a PR for typeshed.

**Comment by li-dan** [comment]
Reading the typeshed contribution guidelines, I saw this:

> Imports in stubs are considered private (not part of the exported API)
> unless:
> * they use the form ``from library import name as name`` (sic, using
>   explicit ``as`` even if the name stays the same); or
> * they use the form ``from library import *`` which means all names
>   from that library are exported.

I wonder if that's the reason that `import ConfigParser as configparser` doesn't work. Anyway, my proposed solution is consistent with what the guidelines say about imports in stubs, since I'm proposing to use ``from library import *``.

**Comment by gvanrossum** [comment]
Well, without `--ignore-missing-imports`, file1.py always gives the error:
```
file1.py:1: error: No library stub file for module 'six.moves.configparser'
file1.py:1: note: (Stub files are from https://github.com/python/typeshed)
```
The reason is that `configparser` is not a submodule of `six.moves`, which would be required to make `import six.moves.configparser` pass.

Now, while that import passes when executed by a Python interpreter, we're on thin ice here: it only works because of the hackery in `six` used to implement `six.moves` efficiently (lazily), and the object `six.moves.configparser` is just an alias for the toplevel module `configparser` (in PY3) or `ConfigParser` (in PY2).

If I wrote less hackish code to define a module alias, an import of that form would not work -- it would be necessary to write `from six.moves import configparser`.

OTOH, the `six` documentation states that e.g. `from six.moves.cPickle import loads` works, so we probably have only two options: either create all the needed subpackages of `six.moves` (and there are a lot of those!), populating each one of them with `from <real_module> import *`, or write a mypy plugin that uses the list of renames from the actual `six` source code to somehow make it work without all those subpackages.
